interpreter/cling/lib/Interpreter/DefinitionShadower.cpp

```diff
@@ -149,14 +160,18 @@ namespace cling {
   ASTTransformer::Result DefinitionShadower::Transform(Decl* D) {
     Transaction *T = getTransaction();
     const CompilationOptions &CO = T->getCompilationOpts();
-    // Global declarations whose origin is the Cling prompt are subject to be
-    // nested in a `__cling_N5' namespace.
+
+    // For variable templates, Transform() is invoked with a VarDecl; get the
+    // corresponding VarTemplateDecl.
+    if (auto VD = dyn_cast<VarDecl>(D))
+      if (auto VTD = VD->getDescribedVarTemplate())
+        D = VTD;
+
+    // Disable definition shadowing for some specific cases.
     if (!CO.EnableShadowing
         || D->getLexicalDeclContext() != m_TU || D->isInvalidDecl()
-        || isa<UsingDirectiveDecl>(D) || isa<UsingDecl>(D)
-        // FIXME: NamespaceDecl requires additional processing (TBD)
-        || isa<NamespaceDecl>(D)
-        || (isa<FunctionDecl>(D) && cast<FunctionDecl>(D)->isTemplateInstantiation())
+        || isa<UsingDirectiveDecl>(D) || isa<UsingDecl>(D) || isa<NamespaceDecl>(D)
+        || isInstantiationOrSpecialization(D)
         || !typedInClingPrompt(FullSourceLoc{D->getLocation(),
                                              m_Context.getSourceManager()}))
       return Result(D, true);
```

Comment:
Why is the second param `true` - IIRC this signals whether there was a replacement or not, and this seems to be the "no replacement" case?

Reply:
According to DeclCollector.cpp:140, the second argument is used to signal success/error condition.